pallets/manta-sbt/src/lib.rs

```diff
@@ -0,0 +1,256 @@
+pub type BalanceOf<T> =
+    <<T as Config>::Currency as Currency<<T as frame_system::Config>::AccountId>>::Balance;
+
+/// MantaSBT Pallet
+#[frame_support::pallet]
+pub mod pallet {
+    use super::*;
+
+    const STORAGE_VERSION: StorageVersion = StorageVersion::new(1);
+
+    /// Pallet
+    #[pallet::pallet]
+    #[pallet::generate_store(pub(super) trait Store)]
+    #[pallet::storage_version(STORAGE_VERSION)]
+    pub struct Pallet<T>(_);
+
+    /// The module configuration trait.
+    #[pallet::config]
+    pub trait Config: frame_system::Config {
+        /// The overarching event type.
+        type Event: From<Event<Self>> + IsType<<Self as frame_system::Config>::Event>;
+
+        /// Weight information for extrinsics in this pallet.
+        type WeightInfo: WeightInfo;
+
+        /// Balance Type
+        type Balance: Default + Member + Parameter + TypeInfo;
+
+        /// The currency mechanism.
+        type Currency: ReservableCurrency<Self::AccountId>;
+
+        /// Pallet ID
+        type PalletId: Get<PalletId>;
+
+        /// Number of unique Asset Ids reserved per `reserve_sbt` call, cannot be zero
+        #[pallet::constant]
+        type MintsPerReserve: Get<u16>;
+
+        /// Price to reserve Asset Ids
+        type ReservePrice: Get<BalanceOf<Self>>;
+
+        /// Private Ledger to Post to
+        type Ledger: PostToLedger<Self::AccountId>;
+
+        /// Register SBTs to `AssetRegistry`
+        type UpdateMetadata: UpdateMetadata<StandardAssetId, Self::Balance>;
+    }
+
+    /// Whitelists accounts to be able to mint SBTs with designated `StandardAssetId`
+    #[pallet::storage]
+    pub type ReservedIds<T: Config> = StorageMap<
+        _,
+        Blake2_128Concat,
+        T::AccountId,
+        (StandardAssetId, StandardAssetId),
+        OptionQuery,
+    >;
+
+    #[pallet::call]
+    impl<T: Config> Pallet<T> {
+        /// Mints a zkSBT
+        #[pallet::call_index(0)]
+        #[pallet::weight(<T as pallet::Config>::WeightInfo::to_private())]
+        #[transactional]
+        pub fn to_private(
+            origin: OriginFor<T>,
+            post: TransferPost,
+            metadata: BoundedVec<u8, SbtBound>,
+        ) -> DispatchResultWithPostInfo {
+            let origin = ensure_signed(origin)?;
+            // Checks that it is indeed a to_private post
+            ensure!(
+                post.sources.len() == 1
+                    && post.sender_posts.is_empty()
+                    && post.receiver_posts.len() == 1
+                    && post.sinks.is_empty(),
+                Error::<T>::NoSenderLedger
+            );
+
+            let (start_id, end_id) =
+                ReservedIds::<T>::get(&origin).ok_or(Error::<T>::NotReserved)?;
+            let asset_id: StandardAssetId = post
+                .asset_id
+                .map(id_from_field)
+                .ok_or(Error::<T>::InvalidAssetId)?
+                .ok_or(Error::<T>::InvalidAssetId)?;
```

Comment:
duplicate error?

Reply:
yeah I think flatten here is better.

One Error is if `TransferPost` is None for `asset_id` field and other is if `id_from_field` fails